File: numi_app/lib/models/trip_map.dart
import 'dart:math' as math;
import 'package:latlong2/latlong.dart';
import 'trip_plan.dart';

LatLng? mapPoint(String latitude, String longitude) {
  final lat = double.tryParse(latitude), lng = double.tryParse(longitude);
  if (lat == null ||
      lng == null ||
      !lat.isFinite ||
      !lng.isFinite ||
      lat.abs() > 90 ||
      lng.abs() > 180) {
    return null;
  }
  return LatLng(lat, lng);
}

LatLng? googleMapPoint(String url) {
  if (linkPlatform(url) != 'Google Maps') return null;
  final uri = safeExternalLink(url);
  if (uri == null || uri.path.contains('/maps/dir/')) return null;
  String raw;
  try {
    raw = Uri.decodeFull('${uri.path}?${uri.query}');
  } on FormatException {
    return null;
  }
  final matches = RegExp(r'!3d(-?\d+(?:\.\d+)?)!4d(-?\d+(?:\.\d+)?)')
      .allMatches(raw)
      .toList();
  if (matches.length == 1) {
    return mapPoint(matches.single[1]!, matches.single[2]!);
  }
  if (matches.length > 1) return null;
  for (final key in ['query', 'q']) {
    final m =
        RegExp(r'^(?:loc:)?\s*(-?\d+(?:\.\d+)?)\s*,\s*(-?\d+(?:\.\d+)?)\s*$')
            .firstMatch(uri.queryParameters[key] ?? '');
    if (m != null) return mapPoint(m[1]!, m[2]!);
  }
  return null; // @lat,lng and ll= are camera positions, not place pins.
}

class TripMapStop {
  final PlanItem item, location;
  final bool arrival;
  final String title, day, time, address;
  final LatLng? point;
  const TripMapStop(
      {required this.item,
      required this.location,
      required this.arrival,
      required this.title,
      required this.day,
      required this.time,
      required this.address,
      required this.point});
  String get key => '${item.id}-${arrival ? 'arrival' : 'place'}';
  String get latitudeField => arrival ? 'endLatitude' : 'latitude';
  String get longitudeField => arrival ? 'endLongitude' : 'longitude';
  bool get saved => item.kind == 'place';
  String? get googleLink => location.links
      .where((l) =>
          l.platform == 'Google Maps' && (l.label == 'Arrival map') == arrival)
      .firstOrNull
      ?.url;
}

List<TripMapStop> tripMapStops(TripPlan plan,
    {String day = '',
    String destination = '',
    String person = '',
    bool savedPlaces = true}) {
  final stops = <TripMapStop>[];
  for (final item in plan.items) {
    if (item.cancelled ||
        !['place', 'activity', 'booking'].contains(item.kind) ||
        !plan.matchesDestination(item, destination)) {
      continue;
    }
    if (item.kind != 'place' && !plan.matchesPerson(item, person)) continue;
    if (item.kind == 'place' && !savedPlaces) continue;
    if (item.kind == 'booking' &&
        item['category'] == 'No accommodation needed') {
      continue;
    }
    final location = plan.find(item['placeId']) ?? item;
    void add(bool arrival) {
      final date = arrival ? item['endDate'] : item['date'];
      if (day.isNotEmpty && item.kind != 'place') {
        if (item.isStay) {
          if (item['date'].compareTo(day) > 0 ||
              item['endDate'].compareTo(day) < 0) {
            return;
          }
        } else if (date != day) {
          return;
        }
      }
      final lat = arrival ? location['endLatitude'] : location['latitude'];
      final lng = arrival ? location['endLongitude'] : location['longitude'];
      var point = mapPoint(lat, lng);
      if (point == null) {
        for (final link in location.links) {
          if ((link.label == 'Arrival map') != arrival) continue;
          point = googleMapPoint(link.url);
          if (point != null) break;
        }
      }
      stops.add(TripMapStop(
          item: item,
          location: location,
          arrival: arrival,
          title: '${plan.itemTitle(item)}${arrival ? ' · Arrival' : ''}',
          day: date,
          time: arrival ? item['endTime'] : item['time'],
          address: arrival ? item['endAddress'] : location['address'],
          point: point));
    }

    add(false);
    if (item.kind == 'booking' &&
        !item.isStay &&
        (item['endAddress'].isNotEmpty || item['endLatitude'].isNotEmpty)) {
      add(true);
    }
  }
  // Don't repeat a saved place already represented by a visible activity.
  final represented =
      stops.where((s) => !s.saved).map((s) => s.location.id).toSet();
  stops.removeWhere((s) => s.saved && represented.contains(s.location.id));
  stops.sort((a, b) {
    if (a.saved != b.saved) return a.saved ? 1 : -1;
    final order = '${a.day} ${a.time.isEmpty ? '99:99' : a.time}'
        .compareTo('${b.day} ${b.time.isEmpty ? '99:99' : b.time}');
    return order != 0
        ? order
        : plan.items.indexOf(a.item).compareTo(plan.items.indexOf(b.item));
  });
  return stops;
}

double mapDistanceKm(LatLng a, LatLng b) {
  double radians(double degrees) => degrees * math.pi / 180;
  final dlat = radians(b.latitude - a.latitude),
      dlng = radians(b.longitude - a.longitude);
  final h = math.pow(math.sin(dlat / 2), 2) +
      math.cos(radians(a.latitude)) *
          math.cos(radians(b.latitude)) *
          math.pow(math.sin(dlng / 2), 2);
  return 6371.0088 * 2 * math.asin(math.sqrt(h.clamp(0, 1)));
}

Uri mapDirections(LatLng from, LatLng to) =>
    Uri.https('www.google.com', '/maps/dir/', {
      'api': '1',
      'origin': '${from.latitude},${from.longitude}',
      'destination': '${to.latitude},${to.longitude}',
    });

File: backend/expenses/templates/expenses/trip_map.js
// Kept inside the collaboration workspace closure: all writes use its revision merge.
let tripMap, mapLayer, mapSignature='', mapSelected='', mapSaved=true, currentStops=[];
function mapPoint(lat,lng){if(lat==null||lng==null||String(lat).trim()===''||String(lng).trim()==='')return null;lat=Number(lat);lng=Number(lng);return Number.isFinite(lat)&&Number.isFinite(lng)&&Math.abs(lat)<=90&&Math.abs(lng)<=180?[lat,lng]:null}
function googlePoint(url){try{const u=new URL(url);if(!/^https?:$/.test(u.protocol)||! /^(?:www\.)?google\.(?:com|[a-z]{2}|com\.[a-z]{2}|co\.[a-z]{2})$/.test(u.hostname)||u.pathname.includes('/maps/dir/'))return null;const raw=decodeURIComponent(u.pathname+u.search),matches=[...raw.matchAll(/!3d(-?\d+(?:\.\d+)?)!4d(-?\d+(?:\.\d+)?)/g)];if(matches.length===1)return mapPoint(matches[0][1],matches[0][2]);if(matches.length>1)return null;for(const k of ['query','q']){const m=(u.searchParams.get(k)||'').match(/^(?:loc:)?\s*(-?\d+(?:\.\d+)?)\s*,\s*(-?\d+(?:\.\d+)?)\s*$/);if(m)return mapPoint(m[1],m[2])}}catch{}return null}
function mapStops(){const stops=[];for(const i of list()){
 if(!['place','activity','booking'].includes(i.kind)||['cancelled','skipped'].includes(i.status)||i.category==='No accommodation needed'||(i.kind==='place'?!mapSaved:!selectedPeople(i)))continue;
 const location=list().find(p=>p.id===i.placeId)||i;
 if(destination&&(i.destinationId||location.destinationId)!==destination&&i.endDestinationId!==destination)continue;
 for(const arrival of [false,...(i.kind==='booking'&&i.category!=='Accommodation'&&(i.endAddress||i.endLatitude)?[true]:[])]){
  const day=(arrival?i.endDate:i.date)||'', time=(arrival?i.endTime:i.time)||'';
  if(selectedDay&&i.kind!=='place'&&(i.category==='Accommodation'?!(i.date<=selectedDay&&i.endDate>=selectedDay):day!==selectedDay))continue;
  const links=(location.links||[]).filter(l=>(l.label==='Arrival map')===arrival);
  const point=mapPoint(location[arrival?'endLatitude':'latitude'],location[arrival?'endLongitude':'longitude'])||links.map(l=>googlePoint(l.url)).find(Boolean)||null;
  stops.push({i,location,arrival,key:i.id+(arrival?'-arrival':'-place'),title:title(i)+(arrival?' · Arrival':''),day,time,point,links});
 }
 }
 const represented=new Set(stops.filter(s=>s.i.kind!=='place').map(s=>s.location.id));
 return stops.filter(s=>s.i.kind!=='place'||!represented.has(s.location.id)).sort((a,b)=>(a.i.kind==='place')-(b.i.kind==='place')||`${a.day} ${a.time||'99:99'}`.localeCompare(`${b.day} ${b.time||'99:99'}`));
}
function renderMap(){
 currentStops=mapStops();const pins=currentStops.filter(s=>s.point),active=pins.find(s=>s.key===mapSelected);
 $('#map-count').textContent=`${pins.length} mapped · ${currentStops.length-pins.length} to locate`;
 $('#map-distance').textContent=active?`Distances from ${active.title} · straight-line, not travel times`:'Select a pin to compare straight-line distances. Open Google Maps for actual routes.';
 $('#map-canvas').hidden=!pins.length;$('#map-empty').hidden=!!pins.length;$('#map-fit').disabled=!pins.length;
 const signature=JSON.stringify([pins.map(s=>[s.key,s.point,s.title]),mapSelected]);
 if(pins.length){
  if(!tripMap){tripMap=L.map('map-canvas',{scrollWheelZoom:false});L.tileLayer('https://tile.openstreetmap.org/{z}/{x}/{y}.png',{maxZoom:19,referrerPolicy:'origin',attribution:'© <a href="https://www.openstreetmap.org/copyright" target="_blank" rel="noopener noreferrer">OpenStreetMap contributors</a>'}).addTo(tripMap).on('tileerror',()=>{$('#map-error').hidden=false});mapLayer=L.layerGroup().addTo(tripMap)}
  requestAnimationFrame(()=>tripMap.invalidateSize());
  if(signature!==mapSignature){const oldPoints=mapSignature?JSON.parse(mapSignature)[0].map(s=>s.slice(0,2)):[];mapLayer.clearLayers();pins.forEach((s,n)=>L.marker(s.point,{title:s.title,icon:L.divIcon({className:'numi-map-pin'+(s.key===mapSelected?' chosen':''),html:String(n+1),iconSize:[32,32],iconAnchor:[16,16]})}).on('click',()=>{mapSelected=s.key;renderMap()}).addTo(mapLayer));if(JSON.stringify(oldPoints)!==JSON.stringify(pins.map(s=>[s.key,s.point])))tripMap.fitBounds(L.latLngBounds(pins.map(s=>s.point)),{padding:[36,36],maxZoom:15});mapSignature=signature}
 }
 $('#cards').innerHTML=currentStops.map((s,index)=>{const pin=pins.indexOf(s),distance=active&&s.point&&active!==s?L.latLng(active.point).distanceTo(s.point)/1000:null;const google=s.point?'https://www.google.com/maps/search/?api=1&query='+encodeURIComponent(s.point.join(',')):'';const route=distance!==null?'https://www.google.com/maps/dir/?api=1&origin='+encodeURIComponent(active.point.join(','))+'&destination='+encodeURIComponent(s.point.join(',')):'';
 return `<article class="card"><div class="row between"><h3>${pin<0?'?':pin+1}. ${esc(s.title)}</h3>${chip(s.i.kind==='place'?'Saved place':personLabel(s.i))}</div><p class="muted">${esc([s.day,s.time].filter(Boolean).join(' '))}${distance!==null?' · '+distance.toFixed(1)+' km away':''}${!s.point?' · Location not set':''}</p><div class="row">${s.point?`<button data-map-select="${index}">Select on map</button><a class="button" href="${esc(google)}" target="_blank" rel="noopener noreferrer">Google Maps ↗</a>${route?`<a class="button" href="${esc(route)}" target="_blank" rel="noopener noreferrer">Directions from selected ↗</a>`:''}`:''}${editable()?`<button data-map-locate="${index}">${s.point?'Change map pin':'Locate'}</button>`:''}</div></article>`}).join('')||'<div class="empty">No places in this view.</div>';
 $('#cards').querySelectorAll('[data-map-select]').forEach(b=>b.onclick=()=>{const s=currentStops[Number(b.dataset.mapSelect)];mapSelected=s.key;renderMap();tripMap.setView(s.point,15)});
 $('#cards').querySelectorAll('[data-map-locate]').forEach(b=>b.onclick=()=>locateMap(currentStops[Number(b.dataset.mapLocate)]));
}
function locateMap(stop){
 const revision=state.revision, original=structuredClone(stop.location);
 const d=modal('Locate '+stop.title,`<form><p>Paste the Google Maps share link for this exact place.</p><label>Google Maps link<input name="url" type="url" required maxlength="3000"></label><p class="error" role="alert"></p><div class="row footer"><button class="primary" type="submit">Locate</button></div></form>`);
 d.querySelector('form').onsubmit=async e=>{e.preventDefault();const b=d.querySelector('[type=submit]');b.disabled=true;try{const url=d.querySelector('input').value.trim();const result=await request('map-preview/',{url});const point=mapPoint(result.latitude,result.longitude);if(!point)throw Error('No exact pin found. Share the place itself, not the map view.');const links=original.links||[];if(!links.some(l=>l.url===url)&&links.length<50)links.push({url,purpose:'Map',label:stop.arrival?'Arrival map':'Google Maps'});await submit([{id:original.id,changes:{[stop.arrival?'endLatitude':'latitude']:String(point[0]),[stop.arrival?'endLongitude':'longitude']:String(point[1]),links}}],revision,()=>d.close())}catch(error){d.querySelector('.error').textContent=error.message}finally{b.disabled=false}};
}
$('#map-saved').onchange=e=>{mapSaved=e.target.checked;renderMap()};
$('#map-fit').onclick=()=>tripMap?.fitBounds(L.latLngBounds(currentStops.filter(s=>s.point).map(s=>s.point)),{padding:[36,36],maxZoom:15});
